File: clojure/p005.clj
; What is the smallest positive number that is evenly divisible by all
; of the numbers from 1 to 20?

(use 'clojure.contrib.math)

; find the prime factors of all the numbers, take the union
(defn prime-factors [number]
  (loop [i 2
         n number
         h {}]
    (if (= n i)
      (assoc h i (inc (get h i 0)))
      (if (zero? (mod n i))
        (recur i (/ n i) (assoc h i (inc (get h i 0))))
        (recur (inc i) n h)))))

(def facs (map prime-factors (range 2 20)))
(println
(reduce
 #(* %1 %2)
 (map #(expt (get % 0) (get % 1))
      (seq (apply merge-with
                  (cons (fn [val1 val2]
                          (max val1 val2))
                        facs)))))
)

File: go/p005.go
// What is the smallest positive number that is evenly divisible by all
// of the numbers from 1 to 20?

// find the prime factors of all the numbers, take the union

package main

import (
	"fmt"
	"math"
)

func factors(x int) map[int]int {
	m := make(map[int]int)
	for i,j := 2,x; i <= j; {
		if j % i == 0  {
			m[i] += 1
			j /= i
		} else {
			i++
		}
	}
	return m
}

func union(x,y map[int]int) map[int]int {
	for k,v := range x {
		y[k] = int(math.Max(float64(v), float64(y[k])))
	}
	return y
}

func main() {
	m := make(map[int]int)
	for i := 2; i < 20; i++ {
		m = union(factors(i), m)
	}
	prod := 1
	for k,v := range m {
		prod *= int(math.Pow(float64(k),float64(v)))
	}
	fmt.Println(prod)
}
